test/replication-luatest/gh_6033_box_promote_demote_test.lua

```diff
@@ -0,0 +1,516 @@
+-- while writing new term to wal.
+g_common.test_wal_interfering_promote = function(g)
+    wal_delay_start(g.server_1)
+    local term = g.server_1:election_term()
+    local f = promote_start(g.server_1)
+    g.server_1:wait_election_term(term + 1)
+
+    g.server_1:exec(function()
+        box.error.injection.set('ERRINJ_RAFT_WAIT_TERM_PERSISTED_DELAY', true)
+    end)
+    promote(g.server_2)
+    wal_delay_end(g.server_1)
+    g.server_1:wait_synchro_queue_term(g.server_2:synchro_queue_term())
+
+    g.server_1:exec(function()
+        box.error.injection.set('ERRINJ_RAFT_WAIT_TERM_PERSISTED_DELAY', false)
+    end)
+
+    local ok, err = fiber_join(g.server_1, f)
+    luatest.assert(not ok and err.code == box.error.INTERFERING_PROMOTE,
+        'Interfering promote fails')
+    wait_sync(g.cluster.servers)
+
+    -- server_1 incremented term and then failed it's promote.
+    -- server_2 isn't leader in new term, but still limbo owner.
+    promote(g.server_1)
+    demote(g.server_1)
+end
+
+-- Demoting should fail if it is interrupted from another server
+-- while writing new term to wal.
+g_common.test_wal_interfering_demote = function(g)
+    promote(g.server_1)
+    wait_sync(g.cluster.servers)
+
+    wal_delay_start(g.server_1)
+    local term = g.server_1:election_term()
+    local f = demote_start(g.server_1)
+    g.server_1:wait_election_term(term + 1)
+
+    g.server_1:exec(function()
+        box.error.injection.set('ERRINJ_RAFT_WAIT_TERM_PERSISTED_DELAY', true)
+    end)
+    promote(g.server_2)
+    wal_delay_end(g.server_1)
+    g.server_1:wait_synchro_queue_term(g.server_2:synchro_queue_term())
+
+    g.server_1:exec(function()
+        box.error.injection.set('ERRINJ_RAFT_WAIT_TERM_PERSISTED_DELAY', false)
+    end)
+
+    local ok, err = fiber_join(g.server_1, f)
+    luatest.assert(not ok and err.code == box.error.INTERFERING_PROMOTE,
+        'Interfering demote fails')
+    wait_sync(g.cluster.servers)
+
+    -- server_1 incremented term and then failed it's demote.
+    -- server_2 isn't leader in new term, but still limbo owner.
+    promote(g.server_1)
+    demote(g.server_1)
+end
+
+-- Promoting should fail if it is interrupted from another server
+-- while waiting for synchro queue being emptied.
+g_common.test_limbo_full_interfering_promote = function(g)
+    -- Need 3 servers for this test:
+    -- server_1 will try to promote with filled synchro queue,
+    -- server_2 will interrupt server_1, while server_3 is leader
+    local box_cfg = table.copy(g.box_cfg)
+    box_cfg.replication = {
+        server.build_instance_uri('server_1'),
+        server.build_instance_uri('server_2'),
+        server.build_instance_uri('server_3'),
+    }
+
+    local server_3 = g.cluster:build_server(
+        {alias = 'server_3', box_cfg = g.box_cfg})
+    server_3:start()
+    box_cfg_update(g.cluster.servers, box_cfg)
+
+    promote(server_3)
+    wait_sync({g.server_1, g.server_2, server_3})
```

Comment:
You don't need this `wait_sync()` if you have another one below anyway.

Reply:
Fixed.